src/arc.rs

```diff
@@ -223,6 +341,206 @@ impl Mul<Arc> for Affine {
+    let mut r = accuracy * 4. * e.powi(-6);
+
+    loop {
+        if 1. <= r * a.powi(6) * a.sqrt() * (1. - e / a) {
+            break;
+        }
+
+        let lambda = (x * y).sqrt() + (x * z).sqrt() + (y * z).sqrt();
+        a = (a + lambda) / 4.;
+        x = (x + lambda) / 4.;
+        y = (y + lambda) / 4.;
+        z = (z + lambda) / 4.;
+
+        r *= 4f64.powi(6);
+        e /= 4.;
+    }
+
+    let x = 1. - x / a;
+    let y = 1. - y / a;
+    let z = -x - y;
+
+    let e2 = x * y - z.powi(2);
+    let e3 = x * y * z;
+
+    (1. + (-1. / 10. * e2 + 1. / 14. * e3 + 1. / 24. * e2.powi(2) - 3. / 44. * e2 * e3)) / a.sqrt()
+}
+
+/// Approximation of the Carlson RD function as defined in "Numerical computation of real or
+/// complex elliptic integrals" (Carlson, Bille C.): <https://arxiv.org/abs/math/9409227v1>
+///
+/// RD = 3/2 ∫ 1 / ( sqrt(t+x) sqrt(t+y) (t+z)^(3/2) ) dt from 0 to inf
+fn carlson_rd(accuracy: f64, x: f64, y: f64, z: f64) -> f64 {
+    // At most one of (x, y) may be 0, z must be nonzero.
+    debug_assert!(z != 0.);
+    debug_assert!(x != 0. || y != 0.);
+
+    // As above for RF, find the absolute upper error bound rather than a relative one, the
+    // derivation of which is along the same lines.
+    //
+    // Again,
+    //
+    // X_n = (a_0 - x_0) / (4^n a_n)
+    // (and the same for variables (Y,y), (Z,z)).
+    //
+    // From "Computing Elliptic Integrals by Duplication" we have
+    //
+    // |err_n| < 4^-n a_n^(-3/2) 3 epsilon_n^6 / (1 - epsilon_n)^(3/2)
+    // with epsilon_n = max(X_n, Y_n, Z_n)
+    //                = max(a_0 - x_0, a_0 - y_0, a_0 - z_0) / (4^n a_n).
+    //
+    // Define e_0 = max(a_0 - x_0, a_0 - y_0, a_0 - z_0). Rewriting for ease of computation,
+    //
+    // |err_n| < 4^-n a_n^(-3/2) 3 epsilon_n^6 / (1 - epsilon_n)^(3/2)
+    //         = 4^-n a_n^(-3/2) 3 e_0^6 / 4^(6n) a_n^6 / (1 - epsilon_n)^(3/2)
+    // -> |err_n| 4^(7n) a_n^(3/2) a_n^6 / (3 e_0^6) < 1 / (1 - epsilon_n)^(3/2)
+    // -> |err_n| 4^(7n) a_n^(3/2) a_n^6 (1/3) / e_0^6 < (1 / 1 - epsilon_n)^(3/2),
+    // raise to the power 2/3,
+    // -> |err_n|^(2/3) 4^(14/3 n) a_n a_n^4 (1/3)^(2/3) / e_0^4 < 1 / (1 - epsilon_n)
+    // -> |err_n|^(2/3) 4^(14/3 n) a_n^5 (1/3)^(2/3) / e_0^4 (1 - epsilon_n) < 1
+    //
+    // That means, to reach an error upper bound of `accuracy`, iterate until
+    // 1 <= accuracy^(2/3) 4^(14/3 n) a_n^5 (1/3)^(2/3) / e_0^4 (1 - epsilon)
+
+    let mut x = x;
+    let mut y = y;
+    let mut z = z;
+
+    let a0 = (x + y + 3. * z) * (1. / 5.);
+    let mut a = a0;
+
+    let mut sum = 0.;
+    let mut mul = 1.;
+
+    // These are partial terms of the inequality derived above. The multiply by (powers of) 4 are
+    // performed per iteration for computational efficiency.
+    let mut e = a - x.min(y).min(z);
+    let mut r = (accuracy / 3.).powf(2. / 3.) * e.powi(-4);
+
+    loop {
+        if 1. <= r * a.powi(5) * (1. - e / a) {
+            break;
+        }
```

Comment:
This bound checking is a non-trivial amount of the operations per iteration, and while I think this is fine, there's an interesting alternative. Iterating to a relative error is cheaper (the formula is in "Numerical computation of real or complex elliptic integrals"), perhaps calculating a conservative upper bound before the loop starts and basing a relative error bound on that is also a possibility.